### PR TITLE
resolves #215: Ability to wait and notify barriers from test framework

### LINKED ISSUE #215: Ability to wait and notify barriers from test framework.
In some use cases we have seen people open up a separate tcp channel in order to notify and await barriers in the script.  This is a hack and we should allow for a standard mechanism for a user to do this.

### PR DESCRIPTION
**Pull request by dpwspoon**
Allows 
```
    @Specification("test.barrier.passing.from.test.framework")
    @Test
    public void testPassingBarriers() throws Exception {
        k3po.awaitBarrier("NOTIFYING_BARRIER");
        k3po.notifyBarrier("AWAITING_BARRIER");
        k3po.finish();
    }
```

Adds `NOTIFY` and `AWAIT` commands to the control protocol, and the `NOTIFIED`, each of which has a header `barrier:barrierName`.  

Additionally, the list of barriers is returned on the `PREPARED` event, by adding multiple `barrier:barrierNames` headers.  This was intended to be backwards compatible as control protocols should ignore unrecognized headers.  However, the control used by junit was not following that and required a fix.  Based on that I recommend bumping the patch version on the next release.

### CODE REVIEW

**Comment by dpwspoon** [comment]
resolves #215

**Comment by dpwspoon** [comment]
@jfallows or @jitsni can you review when you get a chance.  Thanks

File: control/src/main/java/org/kaazing/k3po/control/internal/Control.java [hunk 1]
```diff
@@ -247,8 +281,11 @@ private PreparedEvent readPreparedEvent() throws IOException {
                 case "name":
                     // compatibility
                     break;
+                case "barrier":
+                    prepared.getBarriers().add(headerValue);
+                    break;
                 default:
-                    throw new IllegalStateException("Unrecognized event header: " + headerName);
+                    // NOP allow unrecognized headers for future compatibility
```
Comment (jitsni) [inline control/src/main/java/org/kaazing/k3po/control/internal/Control.java]: Do you want log ?

File: control/src/main/java/org/kaazing/k3po/control/internal/Control.java [hunk 2]
```diff
@@ -230,6 +242,28 @@ private void writeCommand(AbortCommand abort) throws IOException, CharacterCodin
         textOut.flush();
     }
 
+    private void writeCommand(NotifyCommand notify) throws IOException, CharacterCodingException {
+        OutputStream bytesOut = connection.getOutputStream();
+        CharsetEncoder encoder = UTF_8.newEncoder();
+        Writer textOut = new OutputStreamWriter(bytesOut, encoder);
+
+        textOut.append("NOTIFY\n");
+        textOut.append(format("barrier:%s\n", notify.getBarrier()));
```
Comment (jitsni) [inline control/src/main/java/org/kaazing/k3po/control/internal/Control.java]: I haven't seen the control protocol, wouldn't it be better if it looks like "NOTIFY barrier:foo"

File: control/src/main/java/org/kaazing/k3po/control/internal/command/NotifyCommand.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getKind());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return o == this || o instanceof AwaitCommand && equalTo((AwaitCommand) o);
```
Comment (jitsni) [inline control/src/main/java/org/kaazing/k3po/control/internal/command/NotifyCommand.java]: copy paste error ? o instanceof NotifyCommand  and cast to NotifyCommand

File: control/src/main/java/org/kaazing/k3po/control/internal/event/NotifiedEvent.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getKind(), getBarrier());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return o == this || o instanceof PreparedEvent && equalTo((NotifiedEvent) o);
```
Comment (jitsni) [inline control/src/main/java/org/kaazing/k3po/control/internal/event/NotifiedEvent.java]: o instanceof NotifiedEvent ??

File: driver/src/main/java/org/kaazing/k3po/driver/internal/Robot.java [hunk 5]
```diff
@@ -93,6 +96,8 @@
 
     private final ChannelHandler closeOnExceptionHandler = new CloseOnExceptionHandler();
 
+    private Map<String, Barrier> barriersByName;
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/Robot.java]: seems like final field

File: driver/src/main/java/org/kaazing/k3po/driver/internal/Robot.java [hunk 6]
```diff
@@ -527,4 +538,24 @@ public void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e) throws
...
+        if (barrier == null) {
+            throw new Exception("Can not notify nonexistant barrier: " + barrierName);
+        }
+        barrier.getFuture().setSuccess();
+    }
+
+    public ChannelFuture awaitBarrier(String barrierName) throws Exception {
+        final Barrier barrier = barriersByName.get(barrierName);
+        if (barrier == null) {
+            throw new Exception("Can not await nonexistant barrier: " + barrierName);
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/Robot.java]: typo: nonexistent ??

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/AwaitMessage.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(getKind());
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return (this == obj) || (obj instanceof ErrorMessage) && equalTo((ErrorMessage) obj);
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/AwaitMessage.java]: obj instanceof AwaitMessage ?? also equalsTo((AwaitMessage) obj)

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/AwaitMessage.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    protected boolean equals(AwaitMessage that) {
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/AwaitMessage.java]: Similar to other classes equalTo(AwaitMessage) ??

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/NotifiedMessage.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(getKind());
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return (this == obj) || (obj instanceof ErrorMessage) && equalTo((ErrorMessage) obj);
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/NotifiedMessage.java]: Same feedback as other classes. check equals() method in the changeset

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/PreparedMessage.java [hunk 10]
```diff
@@ -16,11 +16,18 @@
...
 package org.kaazing.k3po.driver.internal.control;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Objects;
 
 public class PreparedMessage extends ControlMessage {
 
     private String script = "";
+    private List<String> barriers;
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/PreparedMessage.java]: barriers is a final field ??

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlDownstreamHandler.java [hunk 11]
```diff
@@ -40,6 +40,8 @@ public void writeRequested(ChannelHandlerContext ctx, MessageEvent e) throws Exc
         case ERROR:
             writeErrorRequested(ctx, e);
             break;
+        case NOTIFY:
+            writeNotifyRequested(ctx, e);
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlDownstreamHandler.java]: missing break ??

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlEncoder.java [hunk 12]
```diff
@@ -52,6 +54,10 @@ protected Object encode(ChannelHandlerContext ctx, Channel channel, Object messa
                 return encodeErrorMessage(ctx, channel, (ErrorMessage) controlMessage);
             case FINISHED:
                 return encodeFinishedMessage(ctx, channel, (FinishedMessage) controlMessage);
+            case NOTIFY:
+                return encodeNotifyMessage(ctx, channel, (NotifyMessage) controlMessage);
+            case NOTIFIED:
+                return encodiedNotifiedMessage(ctx, channel, (NotifiedMessage) controlMessage);
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlEncoder.java]: typo in method name

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlServerHandler.java [hunk 13]
```diff
@@ -114,14 +117,16 @@ public void prepareReceived(final ChannelHandlerContext ctx, MessageEvent evt) t
                 ClassLoader contextClassLoader = currentThread.getContextClassLoader();
                 try {
                     currentThread.setContextClassLoader(scriptLoader);
-                    prepareFuture = robot.prepare(aggregatedScript.toString());
+                    prepareFuture =
+                            robot.prepare(aggregatedScript.toString());
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlServerHandler.java]: assignment on the same line looks better (just like the old code)

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlServerHandler.java [hunk 14]
```diff
@@ -229,6 +231,43 @@ public void abortReceived(final ChannelHandlerContext ctx, MessageEvent evt) thr
...
+        }
+        robot.notifyBarrier(barrier);
+        final NotifiedMessage notifiedMessaged = new NotifiedMessage();
+        notifiedMessaged.setBarrier(barrier);
+        logger.debug("sending NOTIFIED: " + barrier);
+        ChannelFuture pendingNotify = ctx.getChannel().write(notifiedMessaged);
+        pendingWrites.add(pendingNotify);
+    }
+
+    CopyOnWriteArrayList<ChannelFuture> pendingWrites = new CopyOnWriteArrayList<>();
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlServerHandler.java]: Move this to where other fields are declared

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlServerHandler.java [hunk 15]
```diff
@@ -229,6 +231,43 @@ public void abortReceived(final ChannelHandlerContext ctx, MessageEvent evt) thr
...
 
+    @Override
+    public void notifyReceived(final ChannelHandlerContext ctx, MessageEvent evt) throws Exception {
+        NotifyMessage notifyMessage = (NotifyMessage) evt.getMessage();
+        final String barrier = notifyMessage.getBarrier();
+        if (logger.isDebugEnabled()) {
+            logger.debug("NOTIFY: " + barrier);
+        }
+        robot.notifyBarrier(barrier);
+        final NotifiedMessage notifiedMessaged = new NotifiedMessage();
```
Comment (jitsni) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlServerHandler.java]: typo notifiedMessage ?

File: driver/src/test/java/org/kaazing/k3po/driver/internal/K3poIT.java [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+public class K3poIT {
+    private final K3poTestRule k3po = new K3poTestRule().setScriptRoot("org/kaazing/specification/control");
+
+    private final TestRule timeout = new DisableOnDebug(new Timeout(3, SECONDS));
+
+    private RobotServer robot;
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
```
Comment (jitsni) [inline driver/src/test/java/org/kaazing/k3po/driver/internal/K3poIT.java]: Is the "trhown" used ??

File: driver/src/test/java/org/kaazing/k3po/driver/internal/K3poIT.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+        robot.start();
+    }
+
+    @After
+    public void shutdownRobot() throws Exception {
+        robot.stop();
+    }
+
+    @Rule
+    public final ExpectedException expectedExceptions = ExpectedException.none();
```
Comment (jitsni) [inline driver/src/test/java/org/kaazing/k3po/driver/internal/K3poIT.java]: Do we need this ?

**Comment by jitsni** [comment]
Is there test with connect barrier is notified from test framework ?

**Comment by jitsni** [comment]
Is there a test with same barrier specified multiple times is notified from test framework ?

File: control/src/main/java/org/kaazing/k3po/control/internal/Control.java (hunk 1 above)
Comment (dpwspoon) [inline control/src/main/java/org/kaazing/k3po/control/internal/Control.java]: Not sure how I would log here as this is on the TestFramework side of thing. (Needed to be logged via junit or System.out)

Comment (dpwspoon) [inline control/src/main/java/org/kaazing/k3po/control/internal/Control.java]: I'll mess around with doing default java logging and see how junit deals with that.

File: control/src/main/java/org/kaazing/k3po/control/internal/Control.java (hunk 2 above)
Comment (dpwspoon) [inline control/src/main/java/org/kaazing/k3po/control/internal/Control.java]: The control protocol is of the form 
```
FrameType
header1:value
header2:value
content-lenth: 
content
```

where the general agreed upon approach is unrecognized headers can be ignored.  But the content needs to be agreed upon.  This was the one thing I reviewed with @jfallows and I think this is better

File: control/src/main/java/org/kaazing/k3po/control/internal/command/NotifyCommand.java (hunk 3 above)
Comment (dpwspoon) [inline control/src/main/java/org/kaazing/k3po/control/internal/command/NotifyCommand.java]: thx

File: control/src/main/java/org/kaazing/k3po/control/internal/event/NotifiedEvent.java (hunk 4 above)
Comment (dpwspoon) [inline control/src/main/java/org/kaazing/k3po/control/internal/event/NotifiedEvent.java]: thx

File: driver/src/main/java/org/kaazing/k3po/driver/internal/Robot.java (hunk 5 above)
Comment (dpwspoon) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/Robot.java]: ok, made it a ConcurrentMap that got passed into constructor of GenerateConfigurationVisitor.  Like it better then it was

File: driver/src/main/java/org/kaazing/k3po/driver/internal/Robot.java (hunk 6 above)
Comment (dpwspoon) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/Robot.java]: changes to "Can not notify a barrier that does not exist in the script"

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlServerHandler.java (hunk 14 above)
Comment (dpwspoon) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlServerHandler.java]: Oops, I don't think the pending writes are needed, it was supposed to be removed.  Thanks

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlServerHandler.java (hunk 15 above)
Comment (dpwspoon) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlServerHandler.java]: I think thats correct, the control handler notifies the barrier when it receives a notify message and then sends a notified response.  But I'll refactor so it reuses the same logic as the await so its more readable :-)  Thanks

File: driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlDownstreamHandler.java (hunk 11 above)
Comment (dpwspoon) [inline driver/src/main/java/org/kaazing/k3po/driver/internal/control/handler/ControlDownstreamHandler.java]: fixed

**Comment by dpwspoon** [comment]
> Is there test with connect barrier is notified from test framework ?
No I'll add one under the connect on barrier tests

> Is there a test with same barrier specified multiple times is notified from test framework ?
No, but I don't have any special logic for this (like I don't throw a error or warning).  All that would happen is the client would be notified multiple times.  I don't think its necessarily beneficial to do so, but what do you think?

**Comment by dpwspoon** [comment]
@jitsni Ready for another round.  I added the two other tests but I'm not sure if the `same barrier specified multiple times is notified` test is needed or whether we should special case any behavior for it.  Let me know if you think we should

**Comment by jitsni** [comment]
```
connect await WRITE_DONE
connect file:target/non.existing.file.dat
        option mode r
connected

....

connect await WRITE_DONE
connect file:target/non.existing.file.dat
        option mode r
connected

....
```

junit test:

```
k3po.notifyBarrier("WRITE_DONE")
```

Isn't the above a valid test case ?

**Comment by dpwspoon** [comment]
@jitsni  Oh I see, added test thanks

**Pull request merged**